CanonicalABI.md

```diff
@@ -0,0 +1,976 @@
+  external WebAssembly module is calling an exported function of a WebAssembly
+  module. Here interface types values, matching the type signature of the
+  module's export, are lowered into core wasm values.
+
+* Lifting an exported function call - this is where a WebAssembly module's
+  export converts the core wasm arguments it received into interface types
+  values.
+
+With a canonical ABI it's expected that almost all of the above operations will
+largely be handled by code generators in various languages. For example if
+you're compiling a Rust project to WebAssembly you'd use a code generator to
+generate all the glue necessary to convert from Rust values to core wasm values,
+according to the canonical ABI.
+
+An expected example of the canonical ABI will be WASI and the functions that it
+supports. The vision for how this might looks is:
+
+1. WASI is first defined in terms of interface types. This means all of its
+   functions will be only using types from the interface types specification.
+
+2. When a WebAssembly module, say C or Rust, calls a WASI API then a code
+   generator for the appropriate language will be used to have an API, in that
+   language, which takes types of that language (e.g. Rust structs, C unions,
+   etc) and then converts those values to the canonical ABI (according to this
+   document). The WebAssembly will have a core wasm import of a WASI function
+   which is expected to behave according to this document.
+
+3. On the implementor side of things, a JS host, for example, would use a code
+   generator to generate an object where it takes core wasm values and then
+   hooks up to a JS implementation of WASI. The JS implementation would only use
+   JS objects and wouldn't need to interact with the wasm module/instance itself
+   (due to the shared-nothing nature of interface types).
+
+Eventually with a binary encoding of interface types it can be imagined that
+WebAssembly modules are wrapped in a thin binary encoding which ascribes the
+types of all imports and exports. With these types various interface types
+signatures can be checked at instantiation time to ensure that the stars all
+align and the ABIs are expected to match. In the very immediate future, however,
+a binary encoding is optional because it can simply be understood that a wasm
+module's imports/exports are described with interface types and both sides of
+the boundary are using the same interface types to generate glue code.
+
+## Interface Types AST
+
+The main thing to "stabilize" for interface types is the grammar for the types
+themselves. For the canonical ABI the assumed type grammar of interface types is:
+
+```
+intertype ::= f32 | f64
+            | s8 | u8 | s16 | u16 | s32 | u32 | s64 | u64
+            | char
+            | (list <intertype>)
+            | (record (field <name> <id>? <intertype>)*)
+            | (variant (case <name> <id>? <intertype>?)*)
+            | (shared-handle <resource>)
```

Comment:
I know I said "shared handle" earlier, but I forgot that `shared` is already claimed to mean "can be called/accessed racily from multiple threads" (as in `(memory 1 shared)` and, in the future, `(func shared ...)`).  Instead of introducing a different adjective, I think it's fine to simply call the copyable version of handles `(handle <resource>)`, treating this as the base case, on which later `(unique <resource>)` and `(borrow <resource>)` refine.